### DISCUSSION TITLE
Limit / Track the number of running BackgroundTasks

### DISCUSSION THREAD

**Discussion by Ben-Epstein** [fastapi/fastapi#8666] [Questions] [answered]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from fastapi import FastAPI, BackgroundTasks

app = FastAPI()

def do_nothing():
    print('hi!')

@app.get("/")
def do_nothing(background_tasks: BackgroundTasks):
    background_tasks.add_task(do_nothing)
    return {"no":"thing"}
```

### Description

I'm trying to understand how BackgroundTasks work in practice, using the doc's specified workflow.

The [docs](https://fastapi.tiangolo.com/tutorial/background-tasks/#using-backgroundtasks) say 

> FastAPI will create the object of type BackgroundTasks for you and pass it as that parameter

Does this mean that each new call to the route will create a new BackgroundTasks object? Or that a single one is created and used across all requests? This would help me understand whether or not infinite background tasks can be created and monitored. 

If the latter is the case, based on the class def below, would it be fine practice to create my own BackgroundTasks object so that I can:
1. Know the queue size
2. Ensure only 1 background task at a time is running

```
class BackgroundTasks(BackgroundTask):
    def __init__(self, tasks: typing.Sequence[BackgroundTask] = []):
        self.tasks = list(tasks)

    def add_task(
        self, func: typing.Callable, *args: typing.Any, **kwargs: typing.Any
    ) -> None:
        task = BackgroundTask(func, *args, **kwargs)
        self.tasks.append(task)

    async def __call__(self) -> None:
        for task in self.tasks:
            await task()
```

Thanks!

### Operating System

macOS

### FastAPI Version

0.66.0

### Python Version

Python 3.8.11

**Comment by Ben-Epstein** [comment]
Any ideas / help on this? Still curious

**Comment by raphaelauv** [comment]
Backgroundtask is an helper

Run your own miltithreadpool or MultiprocessPool to have a fine control on it.

https://github.com/encode/starlette/issues/1094

**Comment by YuriiMotov** [comment, marked as answer]
> Does this mean that each new call to the route will create a new BackgroundTasks object? 

Yes, FastAPI creates new instance for every request

As it's said above, create your own background task manager if you need more control. Or use external distributed tasks ques like Celery or ARQ
